src/environment.js

```diff
@@ -1381,17 +1381,7 @@ export class Reference {
     );
     this.base = base;
     this.referencedName = refName;
-    invariant(
-      !(refName instanceof AbstractValue) ||
-        !(
-          refName.mightNotBeString() &&
-          refName.mightNotBeNumber() &&
-          !refName.isSimpleObject() &&
-          // if the base is a simple abstract object but
-          // the refName is not simple, this is also okay
-          (base instanceof AbstractValue && !base.isSimpleObject())
-        )
-    );
```

Comment:
I removed this `invariant` because I was finding it was getting triggered all the time when in combination with the pure wrapper, which isn't right as indicates that our assumptions at this stage are not right? Maybe @hermanventer has a better idea what to do with this invariant block.

Reply:
The invariant basically states that nothing bad can happen if a reference is delayed. The conditions are too restrictive when you make the additional assumptions of being in pure mode. It seems to me that you could add a disjunction here that succeeds if we are in pure mode.

Reply:
A delayed reference might have issues.

Reply:
Adding a check for isInPureMode is almost always a cheat that will lead to incorrect results because there are no assumptions that shouldn’t be covered by havocing. Anything else is an optimization. So if this isn’t safe in non-pure mode it needs to at least havoc something.

I really want to remove the pure mode because it leads to incorrect assumptions about how we can cheat in this mode.

Reply:
What if I check if the base is `ToObject`?